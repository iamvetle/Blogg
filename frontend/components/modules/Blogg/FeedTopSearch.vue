<template>
    <h2 class="mb-10 text-4xl" v-if="showSearchResults" data-test="search-results">
        SÃ¸keresultater for '{{ searchStore.searchPart }}'
    </h2>
</template>

<script setup lang="ts">
/**
 * Component for displaying search results in the feed top section.
 */

const showSearchResults = computed(() => {
    /**
     * Computed property that determines whether to show the TEXT of the search query.
     * It returns true if the searchPart is not null, otherwise false.
     *
     * @returns {boolean} - True if search results should be shown, false otherwise.
     */
    if (searchStore.searchPart !== null) {
        return true;
    }
});

const searchStore = useSearchStore();
</script>

<style scoped></style>
