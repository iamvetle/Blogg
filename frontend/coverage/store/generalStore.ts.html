
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for store/generalStore.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> / <a href="index.html">store</a> generalStore.ts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">95.23% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>40/42</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>1/1</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/1</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">95.23% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>40/42</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a></td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { defineStore } from "pinia";
import { PersonalPostType } from '../typescript/interfaces';
&nbsp;
export const useGeneralStore = defineStore("general", () =&gt; {
&nbsp;
  // Local storage
  const isAuthenticated = ref(false);
  const username = ref&lt;string | null&gt;(null);
&nbsp;
  // Posts
  const posts = ref&lt;ArticlesSnippetsType | null&gt;(null)
&nbsp;
  // Min side
  const theUser = ref&lt;PersonalPostType | null&gt;(null) 
  const personalPosts = ref&lt;ArticlesSnippetsType | null&gt;(null)
&nbsp;
  // Links
  const baseFeedURL = "http://localhost:8888/api/feed/"
  const post_snippets_url = ref&lt;string&gt;("http://localhost:8888/api/feed/")
  const personal_post_snippets_url = ref&lt;string&gt;("http://localhost:8888/api/min-side/posts/")
&nbsp;
  const next_page_link = ref&lt;string | null&gt;(null)
  const previous_page_link = ref&lt;string | null&gt;(null)
  const last_page_link = ref&lt;string | null&gt;(null)
&nbsp;
  // pagnation
  const total_pages_count = ref&lt;number | null&gt;(null);
  const number_of_posts_count = ref&lt;number | null&gt;(null);
  const current_page = ref&lt;number | null&gt;(null);
&nbsp;
  // error pages
  const apiDownError = ref(false);
&nbsp;
  //other
  const search_bar_show = ref(false)
&nbsp;
  const changeAuthenticated = <span class="fstat-no" title="function not covered" >(state: boolean) =&gt; {</span>
<span class="cstat-no" title="statement not covered" >    isAuthenticated.value = state;</span>
<span class="cstat-no" title="statement not covered" >  };</span>
&nbsp;
  return { posts, theUser, search_bar_show, personalPosts, personal_post_snippets_url, current_page, baseFeedURL, post_snippets_url, isAuthenticated, apiDownError, username, total_pages_count, number_of_posts_count, next_page_link, previous_page_link, last_page_link, changeAuthenticated };
});
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-10-03T20:43:48.495Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    