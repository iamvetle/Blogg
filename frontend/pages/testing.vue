<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>