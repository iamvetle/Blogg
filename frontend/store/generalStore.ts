import { defineStore } from "pinia";

export const useGeneralStore = defineStore("general", () => {

  // Local storage
  const isAuthenticated = ref(false);
  const username = ref<string | null>(null);

  // Posts
  const posts = ref<SnippetPostMultipleType | null>(null)

  // [id] user
  const theUser = ref<NormalUserProfileType | null>(null) 

  // Min side
  const personalPosts = ref<SnippetPostMultipleType | null>(null)
  const personalUser = ref<LoggedInUserProfileType | null>(null)

  // Links
  const baseFeedURL = "http://localhost:8888/api/search/"
  const post_snippets_url = ref<string>("http://localhost:8888/api/feed/")
  const personal_post_snippets_url = ref<string>("http://localhost:8888/api/minkonto/posts/")

  const next_page_link = ref<string>("")
  const previous_page_link = ref<string>("")
  const last_page_link = ref<string>("")

  // pagnation
  const total_pages_count = ref<number>(0);
  const number_of_posts_count = ref<number>(0);
  const current_page = ref<number>(1);

  // error pages
  const apiDownError = ref(false);

  //other
  const search_bar_show = ref(false)

  /** Contains names of all the tags possible */
  const allTags = ref<CategoryType[]>([])

  const idArrayOfSavedPosts = ref<number[]>([])

  const changeAuthenticated = (state: boolean) => {
    isAuthenticated.value = state;
  };

  return { posts, allTags, idArrayOfSavedPosts, personalUser, theUser, search_bar_show, personalPosts, personal_post_snippets_url, current_page, baseFeedURL, post_snippets_url, isAuthenticated, apiDownError, username, total_pages_count, number_of_posts_count, next_page_link, previous_page_link, last_page_link, changeAuthenticated };
});
