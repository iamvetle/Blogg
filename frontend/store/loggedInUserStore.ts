import { defineStore } from 'pinia';

export const useLoggedInUserStore = defineStore("Store for storing information about the logged in user", () => {
    
    /**
     * Contains all of the information about the logged-in user
     */
    const loggedInUserProfile = ref<LoggedInUserProfileType | null>(null) // personalUser


    /**
     * The id of all of the posts the logged in user has saved is stored here
     */
    const idArrayOfSavedPosts = ref<number[]>([])

    /**
     * Keeps the username of all the users the logged-in user is following
     */
    const idArrayOfLoggedInUserFollowingUsers = ref<string[]>([])



    return { loggedInUserProfile, idArrayOfSavedPosts, idArrayOfLoggedInUserFollowingUsers }
})

