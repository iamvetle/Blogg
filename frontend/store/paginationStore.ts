import { defineStore } from 'pinia';

export const usePaginationStore = defineStore("Pagination Store", () => {
    /**
     * This store contains links associated pagination.
     * 
     * It is used to switch between pages. They also updated by search and initial load.
     * 
     */

    /**
     * This url gets update frequently. It is supposed to represent the 
     * last url a fetch call for index posts was made to.
     */

    const postStore = usePostStore()

    const activeFetchURL = ref<any>(urls.api.posts.feed)
    
    const next_page = ref<null | string>(null)
    const previous_page = ref<null | string>(null) 

    const countOfPosts = ref<number>(0);

    const currentPageNumber = ref<number>(1);

    const setPagination = () => {
        if (postStore.posts != null) {
    
            countOfPosts.value = postStore.posts.count
            next_page.value = postStore.posts.next
            previous_page.value = postStore.posts.previous
        }
    }
    
    const route = useRoute()
    const router = useRouter()

    /** Adds the current page as a query to the local route url */
    watchEffect(() => {
        router.replace({
            query: {
                ...route.query,
                page:currentPageNumber.value
            }
        })        
    })


    /**
     * resets everything
     */
    const resetStore = () => {
        activeFetchURL.value = urls.api.posts.feed
        next_page.value = ""
        previous_page.value = ""
    
        countOfPosts.value = 0
    
        currentPageNumber.value = 0 
    }

    return { activeFetchURL, resetStore, setPagination, next_page, previous_page, countOfPosts, currentPageNumber }
})