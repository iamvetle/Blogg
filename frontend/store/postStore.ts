import { defineStore } from 'pinia';

export const usePostStore = defineStore("Store for containing posts and related to posts", () => {

    /**
     * All posts that are fetched in the index are stored here. That includes at
     * initial page load. They have the structure of 'SnippetPostMultipleType'
     */

    /** This contains all posts on feed */
    const posts = ref<SnippetPostMultipleType | null>(null)

    /** This has only the posts made by the users the logged in user is following */
    const followingPosts = ref<SnippetPostMultipleType | null>(null)

    /**
     * All of the posts by the logged in user
     */
    const loggedInUserPosts = ref<LoggedInUserMultiplePostType | null>(null)

/**
 * The url that the baseFetchURL initially was before it started getting modified
 */
    const initialFeedBaseFetchURL = urls.api.posts.feed

    /**
     * The URL that api fetches regarding search generally go to
     */
    const baseFetchURL = ref<string>(initialFeedBaseFetchURL) // post_snippets_url
    const baseLoggedInUserPostsURL = ref<string>(urls.api.posts.myuser) // personal_post_snippets_url ? why is this with ref - think i should remove later

    /** 
     * Has all (paginated) tags 
    */
    const allTags = ref<TagType[] | null>(null)
    /**
     * Has all (paginated) tags 
     */
    const allCategories = ref<CategoryType[] | null>(null)

    /**
     * Contains all of the comments of a post
     * ? should I have this here though?
     */
    const allComments = ref<CommentType[] | null>(null)

    const resetStore = () => {
        posts.value = null
        followingPosts.value = null
        loggedInUserPosts.value = null
        allTags.value = null
        allCategories.value = null
        allComments.value = null
    }

    return { posts, resetStore, allComments, followingPosts, allCategories, allTags, loggedInUserPosts, baseFetchURL, baseLoggedInUserPostsURL };
});
