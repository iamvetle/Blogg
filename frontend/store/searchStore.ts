import { defineStore } from 'pinia'

export const useSearchStore = defineStore("search", () => {
    const searchPosts = ref(null)
    const lastSearch = ref("")

    // Pagnation
    const next_page_link = ref<string | null>(null)
    const previous_page_link = ref<string | null>(null)
    const last_page_link = ref<string | null>(null)

    const total_pages_count = ref<number | null>(null);
    const number_of_posts_count = ref<number | null>(null);
    const current_page = ref<number | null>(null);

    return { searchPosts, lastSearch, next_page_link, previous_page_link, last_page_link, total_pages_count, number_of_posts_count, current_page }
})