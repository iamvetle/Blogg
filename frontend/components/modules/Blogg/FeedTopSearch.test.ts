import FeedTopSearch from './FeedTopSearch.vue';
import { VueWrapper, shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

import { useGeneralStore } from '~/store/generalStore';
import { usePostStore } from '~/store/postStore';
import { useLoggedInUserStore } from '~/store/loggedInUserStore';
import { usePaginationStore } from '~/store/paginationStore';
import { useSearchStore } from '~/store/searchStore';

let wrapper: VueWrapper;
let pinia: any = createTestingPinia();


let searchStore
const standardSearchPart = "tester om fisker finnes"

// let generalStore; 
// let postStore; 
// let loggedInUserStore; 
// let paginationStore; 

const factory = () => {
    return shallowMount(FeedTopSearch, {
        global: {
            plugins: [pinia],
            components: {

            },
            mocks: {},
            stubs: {},
        },
        props: {},
        slots: {}
    })
};

describe('Testing the search top of the feed index', () => {
    // });

    beforeEach(() => {
        // generalStore = useGeneralStore(pinia); 
        // postStore = usePostStore(pinia); 
        // loggedInUserStore = useLoggedInUserStore(pinia); 
        // paginationStore = usePaginationStore(pinia); 
        searchStore = useSearchStore()

        searchStore.searchPart = standardSearchPart

    });

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });
    test('Should exist', () => {
        wrapper = factory()

        expect(wrapper.exists()).toBe(true)
    })
    test('Should display the text "søkeresultater" if there is a true search value', () => {
        wrapper = factory()

        const searchResults = wrapper.find("[data-test='search-results']")

        // the element existance
        expect(searchResults.exists()).toBe(true)

        // only the text
        expect(searchResults.text()).toContain("Søkeresultater for")

        // text and search value
        expect(searchResults.text()).toBe(`Søkeresultater for '${standardSearchPart}'`)

    })
    test('Should match snapshot', () => {
        wrapper = factory()
        expect(wrapper.html()).toMatchSnapshot()
    })


});