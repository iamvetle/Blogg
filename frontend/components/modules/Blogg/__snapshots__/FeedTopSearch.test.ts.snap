// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the search top of the feed index > Should match snapshot 1`] = `
"<div>
  <div>
    <h2 class=\\"mb-10 text-4xl\\" data-test=\\"search-results\\"> SÃ¸keresultater for 'tester om fisker finnes' </h2>
  </div>
</div>"
`;
